rllib/algorithms/dt/segmentation_buffer.py

```diff
@@ -0,0 +1,215 @@
+
+
+class SegmentationBuffer:
+    """A minimal replay buffer used by Decision Transformer (DT)
+    to process episodes into max_seq_len length segments and do shuffling.
+    """
+
+    def __init__(
+        self,
+        capacity: int = 20,
+        max_seq_len: int = 20,
+        max_ep_len: int = 1000,
+    ):
+        """
+        Args:
+            capacity: Maximum number of episodes the buffer can store.
+            max_seq_len: Length of segments that are sampled.
+            max_ep_len: Maximum length of episodes added.
+        """
+        self.capacity = capacity
+        self.max_seq_len = max_seq_len
+        self.max_ep_len = max_ep_len
+
+        self._buffer: List[SampleBatch] = []
+
+    def add(self, batch: SampleBatch):
+        """Add a SampleBatch of episodes. Replace if full.
+
+        Args:
+            batch: SampleBatch of full episodes.
+        """
+        episodes = batch.split_by_episode()
+        for episode in episodes:
+            self._add_single(episode)
+
+    def _add_single(self, episode: SampleBatch):
+        # Truncate if episode too long.
+        # Note: sometimes this happens if the dataset shuffles such that the
+        # same episode is concatenated together twice (which is okay).
+        if episode.env_steps() > self.max_ep_len:
+            logger.warning(
+                f"The maximum rollout length is {self.max_ep_len} but we tried to add a"
+                f"rollout of {episode.env_steps()} steps to the SegmentationBuffer. "
+                f"This could be due to incorrect data in the dataset (need fixing), "
+                f"or random shuffling that caused a duplicate rollout (no need to fix)."
+            )
+            episode = episode[: self.max_ep_len]
+
+        if len(self._buffer) < self.capacity:
+            self._buffer.append(episode)
+        else:
+            # TODO: replace proportional to episode length
+            replace_ind = np.random.randint(0, self.capacity)
+            self._buffer[replace_ind] = episode
+
+    def sample(self, batch_size: int) -> SampleBatch:
+        """Sample segments from the buffer.
+
+        Args:
+            batch_size: number of segments to sample.
+
+        Returns:
+            SampleBatch of segments with keys and shape {
+                OBS: [batch_size, max_seq_len, obs_dim],
+                ACTIONS: [batch_size, max_seq_len, act_dim],
+                RETURNS_TO_GO: [batch_size, max_seq_len + 1, 1],
+                T: [batch_size, max_seq_len],
+                ATTENTION_MASKS: [batch_size, max_seq_len],
+            }
+        """
+        samples = [self._sample_single() for _ in range(batch_size)]
+        return concat_samples(samples)
+
+    def _sample_single(self) -> SampleBatch:
+        # TODO: sample proportional to episode length
+        # Sample a random episode from the buffer and then sample a random
+        # segment from that episode.
+        buffer_ind = np.random.randint(0, len(self._buffer))
```

Comment:
this is not a seedable call. Can you instantiate your class with a numpy rng and set the seed of the rng using the seed in the global config?

Reply:
resolved